klippy/extras/frame_expansion_compensation.py

```diff
@@ -0,0 +1,195 @@
+
+    def stats(self, eventtime):
+        return False, '%s: temp=%.1f' % ('Frame', self.last_temp)
+
+    def get_status(self, eventtime):
+        return {
+            'temperature': self.last_temp,
+            'measured_min_temp': self.measured_min,
+            'measured_max_temp': self.measured_max,
+            'current_z_comp': self.z_drift_offset,
+            'frame_ref_temp': self.last_home_temp,
+            'state': self.comp_state
+        }
+
+    def check_eligible(self, endstops):
+        'Check if homed axis is `stepper_z`.'
+        steppers_being_homed = [s.get_name()
+                                for es in endstops
+                                for s in es.get_steppers()]
+        return self.z_stepper_name in steppers_being_homed
+
+    def handle_homing_move_end(self, hmove):
+        'Triggered when Z axis is homed.'
+        if self.check_eligible(hmove.get_mcu_endstops()):
+            self.last_home_temp = self.last_temp
+            self.z_drift_offset = 0.
+
+    def calc_offset(self, pos):
+        'Calculate total linear thermal expansion relative to last homing.'
+        if not self.max_comp_z or pos[2] < self.max_comp_z:
+            # Calculate the extrusion length at homing temperature
+            length_home = (self.frame_z + self.frame_z * ( self.coeff *
+                            (self.last_home_temp - STD_TEMP) ))
+            delta_t = self.last_temp - self.last_home_temp
+
+            # Calculate Z offset, compensates for thermal expansion since homing
+            offset = -1 * (length_home *
+                (self.coeff * delta_t) * self.gantry_factor) * 1E3
+
+            # compute sign (+1 or -1) for maximum offset setting
+            sign = 1 - (offset <= 0)*2
+
+            self.z_drift_offset = min([self.max_offset*sign, offset], key=abs)
+
+        self.update_state()
+
+        # Apply offset
+        new_z = pos[2] + self.z_drift_offset
+        return [pos[0], pos[1], new_z, pos[3]]
+
+    def calc_unoffset(self, pos):
+        'Remove Z-drift offset'
+        unoffset_z = pos[2] - self.z_drift_offset
+        return [pos[0], pos[1], unoffset_z, pos[3]]
+
+    def get_position(self):
+        return self.calc_unoffset(self.next_transform.get_position())
+
+    def move(self, newpos, speed):
+        # don't apply to extrude only moves or when disabled
+        if (newpos[0:2] == self.last_position[0:2]) or not self.comp_enable:
+            self.next_transform.move(newpos, speed)
+        else:
+            corrected_pos = self.calc_offset(newpos)
+            self.next_transform.move(corrected_pos, speed)
+            self.last_position[:] = newpos
+        self.last_position[:] = newpos
+
+    def temperature_callback(self, read_time, temp):
+        'Called everytime the thermistor is read'
+        self.last_temp = temp
+        if temp:
+            self.measured_min = min(self.measured_min, temp)
+            self.measured_max = max(self.measured_max, temp)
```

Comment:
The `if temp` check doesn't make sense - it's always valid (or there is a severe error occurring).  Unless there is a need for it, I don't think this code should be tracking measure_min/measured_max.

Reply:
I'm not running Mainsail or other Moonraker-interfaced frontends, so I can't test this at the moment without some effort, but in so far as I understood min/max tracking was necessary for compatibility. It was a kind contribution from @fsironman necessary to get it running on his system.

Reply:
The frame_expansion_compensation class is more or less a **temperature sensor class** that can apply a Z-offset.
The feature itself will work without tracking measure_min/measured_max yes,
but wont work with any third party (in this case moonrake/mainsail)
for which it makes sense to display the temperature from frame_expansion_compensation in its UI.

Hence the same functions as the generic temperature sensor are included (https://github.com/KevinOConnor/klipper/blob/master/klippy/extras/temperature_sensor.py)

This will allow 3rd party pluging to display the temperature in its UI as if its a generic temperature sensor:
![unknown](https://user-images.githubusercontent.com/6130672/118362219-f24c4400-b58e-11eb-9f5a-d768041b82f5.png)
(below without it, on top proper display with including min/max change)